File: patterns/creational/builder/javascript/src/product.js
export class Product {
  #parts = [];

  /** @returns {readonly string[]} */
  get parts() {
    return this.#parts.slice();
  }

  /** @param {string} part */
  add(part) {
    if (part.trim().length === 0) {
      throw new Error("A product part cannot be empty.");
    }

    this.#parts.push(part);
  }

  listParts() {
    return `Product parts: ${this.#parts.join(", ")}`;
  }
}

File: patterns/creational/builder/csharp/src/Builder/Product.cs
namespace Builder;

public sealed class Product
{
    private readonly List<string> _parts = [];

    public IReadOnlyList<string> Parts => _parts;

    public void Add(string part)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(part);
        _parts.Add(part);
    }

    public string ListParts() => $"Product parts: {string.Join(", ", _parts)}";
}
